### ISSUE TITLE
Unable to use entire JSON column in .where() clause

### ISSUE THREAD

**Issue by edebill** [rails/rails#25463]
### Steps to reproduce

Create an ActiveRecord model with a JSON column in postgres (jsonb), then try to use
that column as part of a where() statement. The only way I can get a match is to use the SQL
placeholder syntax Example.where("numbers = ?", numbers.to_json), instead of either of the
two previous syntaxes that used to work at least as recently a 5.0.0.beta1  Example.where(numbers: numbers) or Example.where(numbers: numbers.to_json).

I built a clean rails 5.0.0.rc1 app with tests that demonstrate this. Just bin/rake test to get a few kinds of failures and a passing test for the workaround I mention above.
[repo with repro case](https://github.com/edebill/repro)

### Expected behavior

I expect that the following will pass
```
  # ActiveRecord thinks it's nested/joined tables
  test "query with hash structure" do
    numbers = { list: [1,2,3] }
    e = Example.create(numbers: numbers)
    a = Example.where(numbers: numbers).first
    assert_equal e, a, "didn't find Example"
  end

  # ActiveRecord double-escapes the string, so it does not match
  test "query with JSON string" do
    numbers = { list: [1,2,3] }
    e = Example.create(numbers: numbers)
    a = Example.where(numbers: numbers.to_json).first
    assert_equal e, a, "didn't find Example"
  end
```

### Actual behavior
In the first case above, where I pass { list: [1,2,3] } I get an invalid SQL error as it interprets it as wanting to join on a table named after my json column.

```
2.3.1 :007 > numbers = { list: [1,2,3] }
 => {:list=>[1, 2, 3]}
2.3.1 :008 > f = Example.where(numbers: numbers).first
  Example Load (5.1ms)  SELECT  "examples".* FROM "examples" WHERE "numbers"."list" IN (1, 2, 3) ORDER BY "examples"."id" ASC LIMIT $1  [["LIMIT", 1]]
ActiveRecord::StatementInvalid: PG::UndefinedTable: ERROR:  missing FROM-clause entry for table "numbers"
LINE 1: SELECT  "examples".* FROM "examples" WHERE "numbers"."list" ...
```

 In the second one above, it generates 
```
f = Example.where(numbers: numbers.to_json).first
  Example Load (0.7ms)  SELECT  "examples".* FROM "examples" WHERE "examples"."numbers" = $1 ORDER BY "examples"."id" ASC LIMIT $2  [["numbers", "\"{\\\"list\\\":[1,2,3]}\""], ["LIMIT", 1]]
```

Which looks like it interprets the JSON string as a string that should be converted to JSON and double-escapes everything.

### System configuration
**Rails version**: 5.0.0.rc1

**Ruby version**: ruby 2.3.1p112 (2016-04-26 revision 54768) [x86_64-darwin15]
        and 2.3.0.p0

**Comment by maclover7** [comment]
Combined tests into a single reproduction script --> https://gist.github.com/maclover7/ee94183eb5d1c281e4c696d124548130

**Comment by devonestes** [comment]
Hey y'all,

So I'd love to help out with trying to fix this bug! It seems to me at first glance like something that's up with the type definition for the jsonb type in the Attributes API. This would be my first attempt at contributing to Rails, so I'm excited, but I'd also really love any guidance folks might have to offer about where to put the tests for this bug (since it's adapter specific, should it go there?), and also where I might be able to find the files to change.

I'll start looking myself, though, and hopefully I can help!

**Comment by rafaelfranca** [comment]
Hey!

Glad you want to help. Your tests should probably go in https://github.com/rails/rails/blob/master/activerecord/test/cases/adapters/postgresql/json_test.rb. There is also a file to json type in mysql.

**Comment by kamipo** [comment]
This jsonb issue is fixed by #25671.
But json type still need to cast explicitly like `SELECT "json_data_type".* FROM "json_data_type" WHERE "json_data_type"."payload" = '{"color":"green"}'::json LIMIT 1`.

```
ActiveRecord::StatementInvalid: PG::UndefinedFunction: ERROR:  operator does not exist: json = unknown
LINE 1: ..."json_data_type" WHERE "json_data_type"."payload" = '{"color...
                                                             ^
HINT:  No operator matches the given name and argument type(s). You might need to add explicit type casts.
: SELECT "json_data_type".* FROM "json_data_type" WHERE "json_data_type"."payload" = '{"color":"green"}' LIMIT 1
```
